File: app/src/main/java/com/jackyli/androidarchitecture/ui/vms/LoginViewModel.java
package com.jackyli.androidarchitecture.ui.vms;

import android.annotation.SuppressLint;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;


import com.jackyli.androidarchitecture.api.ApiService;
import com.jackyli.androidarchitecture.model.Result;
import com.jackyli.androidarchitecture.model.UserInfo;
import com.jackyli.androidarchitecture.repository.db.AppDataBase;
import com.jackyli.androidarchitecture.repository.net.RetrofitManager;

import java.util.List;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class LoginViewModel extends ViewModel {

    public MutableLiveData<Boolean> mLDLogin;
    private final LiveData<List<UserInfo>> listLiveData = new MutableLiveData<>();

    public MutableLiveData<Boolean> getLoginResult() {
        if (mLDLogin == null) {
            mLDLogin = new MutableLiveData<>();
        }
        return mLDLogin;
    }

    public void login(String name, String pass) {
        Call<Result> call = RetrofitManager.get().create(ApiService.class).loginNetwork(name, pass);
        call.enqueue(new Callback<Result>() {
            @Override
            public void onResponse(Call<Result> call, Response<Result> response) {
                Result result = response.body();
                assert result != null;
                if (result.getResult().equals("Success")) {
                    mLDLogin.postValue(true);
                    UserInfo userInfo = new UserInfo();
                    userInfo.setUserName(name);
                    userInfo.setPassWord(pass);
                    if (listLiveData.getValue() != null && listLiveData.getValue().size() > 0) {
                        insertDb(userInfo);
                    }
                } else {
                    mLDLogin.postValue(false);
                }
            }

            @Override
            public void onFailure(Call<Result> call, Throwable t) {
                mLDLogin.postValue(false);
            }
        });
    }

    public LiveData<List<UserInfo>> getUser() {
        return AppDataBase.getInstance().getUserDao().getUsers();
    }

    @SuppressLint("CheckResult")
    public void insertDb(UserInfo userInfo) {
        //可以考虑使用线程池或者rxjava来做
        AppDataBase.getInstance().getUserDao().addUser(userInfo).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).subscribe(integer -> {

        }, throwable -> {

        });
    }
}

File: app/src/main/java/com/jackyli/androidarchitecture/ui/vms/LoginViewModel.kt
package com.jackyli.androidarchitecture.ui.vms

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.jackyli.androidarchitecture.model.UserInfo
import com.jackyli.androidarchitecture.repository.db.AppDataBase
import com.jackyli.androidarchitecture.repository.net.RetrofitManager
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import java.lang.Exception

/**
 *  利用协程来使用调度
 *
 *  GlobalScope是生命周期是process级别的，即使Activity或Fragment已经被销毁，协程仍然在执行。所以需要绑定生命周期。
 *  lifecycleScope只能在Activity、Fragment中使用，会绑定Activity和Fragment的生命周期
 *  viewModelScope只能在ViewModel中使用，绑定ViewModel的生命周期
 */
class LoginViewModel : ViewModel() {

    var liveDataLogin: MutableLiveData<Boolean> = MutableLiveData()
    var liveDataUser: LiveData<List<UserInfo>> = MutableLiveData()

    fun getLoginResult(): MutableLiveData<Boolean> {
        return liveDataLogin
    }

    fun getUserInfoFromDb(): LiveData<List<UserInfo>> {
        liveDataUser = AppDataBase.getInstance().getUserDao().getUsers()
        return liveDataUser
    }

    fun login(name: String, pass: String) {
        viewModelScope.launch(Dispatchers.IO) {
            try {
                val retrofit2 = RetrofitManager()
                val execute = retrofit2.service.loginNetWork(name, pass).execute()
                val result = execute?.body()
                if (result?.result == "Success") {
                    liveDataLogin.postValue(true)

                    if (liveDataUser.value == null || liveDataUser.value?.size == 0) {
                        val userInfo = UserInfo(0, name, pass)
                        insertUser(userInfo)
                    }
                } else {
                    liveDataLogin.postValue(false)
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    fun insertUser(userInfo: UserInfo) {
        viewModelScope.launch(Dispatchers.IO) {
            AppDataBase.getInstance().getUserDao().insertUser(userInfo)
        }
    }
}